### PR TITLE
Use generic for api version to reduce runtime branching

### PR DESCRIPTION
**Pull request by andylokandy**
ref #10938

Signed-off-by: andylokandy <andylokandy@hotmail.com>

- Add a component crate `api_version` which contains a generic trait `APIVersion` for as the const variant of `kvrpcpb::ApiVersion`.
- Migrate `engine_traits::key_prefix` and `engine_traits::raw_value` modules into `APIVersion`.
- Add a macro `match_template_api_version!()` to lift runtime enum `kvrpcpb::ApiVersion` to const `APIVersion`.
- Use `APIVersion` instead of  `kvrpcpb::ApiVersion` in `SSTImporter`, `RawStore`, `RawEncodedSnapshot`, `CompactionFilter`.
- Use `txn_types::Key` instead of `Vec<u8>` for keys of transaction commands in `storage/mod.rs`
- Add `check_api_version_ranges` in  `storage/mod.rs`.

### What is changed and how it works?

Ref #10938

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
none
```

### CODE REVIEW

File: components/api_version/src/api_v1ttl.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        })
+    }
+
+    fn encode_raw_value(value: RawValue<&[u8]>) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(value.user_value.len() + number::U64_SIZE);
+        buf.extend_from_slice(value.user_value);
+        buf.encode_u64(value.expire_ts.unwrap_or(0)).unwrap();
+        buf
+    }
+    fn encode_raw_value_owned(mut value: RawValue<Vec<u8>>) -> Vec<u8> {
```
Comment (zhongzc) [inline components/api_version/src/api_v1ttl.rs:48]:
```suggestion

    fn encode_raw_value_owned(mut value: RawValue<Vec<u8>>) -> Vec<u8> {
```

File: components/sst_importer/src/sst_writer.rs [hunk 2]
```diff
@@ -182,29 +185,35 @@ impl<E: KvEngine> RawSSTWriter<E> {
...
+                                self.put(m.get_key(), &API::encode_raw_value(value), PairOp::Put)?;
+                            }
+                            PairOp::Delete => {
+                                self.put(m.get_key(), &[], PairOp::Delete)?;
+                            }
+                        };
+                    }
                 }
-            };
-        }
```
Comment (zhongzc) [inline components/sst_importer/src/sst_writer.rs:207]: back to original code

**Review by zhongzc — APPROVED** [review]

**Review by Connor1996 — COMMENTED** [review]
rest LGTM

File: components/api_version/src/api_v2.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    fn parse_key_mode(key: &[u8]) -> KeyMode {
+        if key.is_empty() {
+            return KeyMode::Unknown;
+        }
+
+        match key[0] {
+            RAW_KEY_PREFIX => KeyMode::Raw,
+            TXN_KEY_PREFIX => KeyMode::Txn,
+            b'm' | b't' => KeyMode::TiDB,
```
[the remark this answers is not in the document]
Comment (andylokandy) [reply components/api_version/src/api_v2.rs]: done

**Review by Connor1996 — APPROVED** [review]
LGTM

**Comment by andylokandy** [comment]
https://github.com/tikv/tikv/pull/11664 would be blocked by this PR. I'm going to merge it first.

**Pull request merged**